ipfsconn/ipfshttp/ipfshttp.go

```diff
@@ -323,9 +357,166 @@ func (ipfs *Connector) pinLsHandler(w http.ResponseWriter, r *http.Request) {
 		}
 	}
 
-	respBytes, _ := json.Marshal(pinLs)
+	resBytes, _ := json.Marshal(pinLs)
 	w.WriteHeader(http.StatusOK)
-	w.Write(respBytes)
+	w.Write(resBytes)
+}
+
+func (ipfs *Connector) addHandler(w http.ResponseWriter, r *http.Request) {
+	// Handle some request options
+	q := r.URL.Query()
+	// Remember if the user does not want cluster/ipfs to pin
+	doNotPin := q.Get("pin") == "false"
+	// make sure the local peer does not pin.
+	// Cluster will decide where to pin based on metrics and current
+	// allocations.
+	q.Set("pin", "false")
+	// do not send progress updates. They have no use since we need to
+	// wait until the end to tell cluster to pin.
+	q.Set("progress", "false")
+	r.URL.RawQuery = q.Encode()
+
+	res, err := ipfs.proxyRequest(r)
+	if err != nil {
+		http.Error(w, "error forwarding request: "+err.Error(), 500)
+		return
+	}
+	defer res.Body.Close()
+
+	// Shortcut some cases where there is nothing else to do
+	if scode := res.StatusCode; scode != http.StatusOK {
+		logger.Warningf("proxy /add request returned %d", scode)
+		ipfs.proxyResponse(w, res, res.Body)
+		return
+	}
+
+	if doNotPin {
+		logger.Debug("proxy /add requests has pin==false")
+		ipfs.proxyResponse(w, res, res.Body)
+		return
+	}
+
+	// The ipfs add response is a streaming-like body where
+	// { "Name" : "filename", "Hash": "cid" } objects are provided
+	// for every added object.
+
+	// We will need to re-read the response to re-play to the client
+	// at the end, therefore we make a copy in bodyCopy while decoding.
+	bodyCopy := new(bytes.Buffer)
+	bodyReader := io.TeeReader(res.Body, bodyCopy)
+
+	ipfsAddResps := []ipfsAddResp{}
+	dec := json.NewDecoder(bodyReader)
+	for dec.More() {
+		var addResp ipfsAddResp
+		err := dec.Decode(&addResp)
+		if err != nil {
+			http.Error(w, "error decoding response: "+err.Error(), 502)
+			return
+		}
+		ipfsAddResps = append(ipfsAddResps, addResp)
+	}
+
+	if len(ipfsAddResps) == 0 {
+		logger.Warning("proxy /add request response was OK but empty")
+		ipfs.proxyResponse(w, res, bodyCopy)
+		return
+	}
+
+	// An ipfs add call can add multiple files and pin multiple items.
+	// ipfs api is not perfectly behaved here (i.e. when passing in two
+	// directories to pin). There is no easy way for sure what has
+	// been pinned recursively and what not.
+	// Usually when pinning a directory, the recursive pin comes last.
+	// But we may just be pinning different files and no directories.
+	// In that case, we need to recursively pin them separately.
+	// Decide recursivePins takes a conservative approach. It
+	// works on the regular use-cases. Otherwise, it might pin
+	// more things than it should.
+	pinHashes := decideRecursivePins(ipfsAddResps, r.URL.Query())
+
+	logger.Debugf("proxy /add request and will pin %s", pinHashes)
+	for _, pin := range pinHashes {
```

Comment:
Does this mean that it will separately pin all files in case of big directory?

Is that how cluster works?